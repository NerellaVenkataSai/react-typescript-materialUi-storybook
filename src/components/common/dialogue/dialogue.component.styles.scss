.dialogue-component {
	.dialogue-header {
		padding-top: 10px;
		padding-bottom: 10px;
	}	
	.MuiDialogTitle-root {
		padding: 0 14px;
		.close-button {
			position: absolute;
			right: 0;
			padding-top: 4px;
		}
	}
}
