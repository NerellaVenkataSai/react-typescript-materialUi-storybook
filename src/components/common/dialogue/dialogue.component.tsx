import './dialogue.component.styles.scss';

import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Grid,
  IconButton,
  PropTypes,
  Typography,
} from '@material-ui/core';
import { DialogProps } from '@material-ui/core/Dialog';
import { Close } from '@material-ui/icons';
import React, { FunctionComponent } from 'react';

export type DialogAction = {
  action: 'Save' | 'Update' | 'Cancel';
  text: string;
  color?: PropTypes.Color;
  isDisabled?: boolean;
  type?: 'submit' | 'normal';
  formId?: string;
};
export type DialogComponentProps = {
  actions?: DialogAction[];
  title: string;
  open: boolean;
  body?: string;
  className?: string;
  dividers?: boolean;
  onClose?: () => void;
  onAction?: (action: DialogAction['action'], index: number) => void;
} & DialogProps;

export const DialogComponent: FunctionComponent<DialogComponentProps> = ({
  actions,
  body,
  children,
  className,
  dividers,
  onClose,
  onAction,
  open,
  title,
  ...props
}) => (
  <Dialog
    className={`dialogue-component ${className}`}
    onClose={onClose}
    aria-labelledby={title}
    open={open}
    {...props}
  >
    <DialogTitle>
      <Grid container justify="flex-start" className="dialogue-header">
        <Grid item xs={10}>
          <Typography variant="h6">{title}</Typography>
        </Grid>
        {onClose ? (
          <Grid item xs={2}>
            <IconButton aria-label="Close" className="close-button" onClick={onClose}>
              <Close />
            </IconButton>
          </Grid>
        ) : (
          ''
        )}
      </Grid>
    </DialogTitle>
    <DialogContent dividers={dividers}>
      {children ? children : <DialogContentText>{body}</DialogContentText>}
    </DialogContent>
    <DialogActions>
      {actions?.map(({ action, color, text, isDisabled, type = 'normal', formId }: DialogAction, index: number) => (
        <>
          {type === 'submit' ? (
            <Button
              key={index}
              disabled={isDisabled}
              color={color || 'default'}
              variant="contained"
              // fullWidth
              type="submit"
              form={formId || 'custom-form'}
            >
              {text}
            </Button>
          ) : (
            <Button
              key={index}
              disabled={isDisabled}
              color={color || 'default'}
              variant="contained"
              // fullWidth
              onClick={(e) => {
                e.preventDefault();
                onAction && onAction(action, index);
              }}
            >
              {text}
            </Button>
          )}
        </>
      ))}
    </DialogActions>
  </Dialog>
);
