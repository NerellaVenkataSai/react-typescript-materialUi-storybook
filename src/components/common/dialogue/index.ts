export * from './dialogue.component';
export * from './dialogue-hook';
