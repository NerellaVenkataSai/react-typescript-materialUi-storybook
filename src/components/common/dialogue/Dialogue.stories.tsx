import { Button } from '@material-ui/core';
import { Meta, Story } from '@storybook/react';
import { DialogComponent, DialogComponentProps, useDialogueHook } from 'components/common/dialogue';
import React from 'react';

export default {
  title: 'Dialogue',
  //   argTypes: { onClick: { action: 'clicked' } },
  component: DialogComponent,
  parameters: {
    docs: {
      source: {
        type: 'code',
      },
    },
  },
} as Meta;

export const SimpleDialog: Story<DialogComponentProps> = ({ body, onClose, title, ...args }) => {
  const { open, setOpen } = useDialogueHook();
  return (
    <>
      <Button color="primary" variant="contained" onClick={() => setOpen(!open)} size="medium">
        Click Here To Open Dialog
      </Button>
      <DialogComponent
        {...args}
        className="personalInfo-dialog-component"
        color="primary"
        open={open}
        /**
         * kept this or condition just to take dynamic args from stroybook UI
         */
        title={title || 'Simple Dialogue'}
        body={body || 'Dialog without close action or icon, please do a page refresh or click on canvas'}
      />
    </>
  );
};

export const DialogWithCloseIcon: Story<DialogComponentProps> = ({ body, title, ...args }) => {
  const { open, setOpen } = useDialogueHook();
  return (
    <>
      <Button color="primary" variant="contained" onClick={() => setOpen(!open)} size="medium">
        Click Here To Open Dialog
      </Button>
      <DialogComponent
        {...args}
        className="personalInfo-dialog-component"
        color="primary"
        open={open}
        onClose={() => setOpen(false)}
        /**
         * kept this or condition just to take dynamic args from stroybook UI
         */
        title={title || 'Closable Dialog'}
        body={body || 'Dialog can be closed by clicking on cross icon or on outside click of Dialog'}
      />
    </>
  );
};

export const DialogWithDividers: Story<DialogComponentProps> = ({ body, title, ...args }) => {
  const { open, setOpen } = useDialogueHook();
  return (
    <>
      <Button color="primary" variant="contained" onClick={() => setOpen(!open)} size="medium">
        Click Here To Open Dialog
      </Button>
      <DialogComponent
        {...args}
        className="personalInfo-dialog-component"
        color="primary"
        open={open}
        onClose={() => setOpen(false)}
        /**
         * kept this or condition just to take dynamic args from stroybook UI
         */
        dividers={true}
        title={title || 'Dialog with dividers'}
        body={body || 'Dialog with header and footer separator'}
      />
    </>
  );
};

export const DialogWithScrollableContent: Story<DialogComponentProps> = ({ body, title, ...args }) => {
  const { open, setOpen } = useDialogueHook();
  return (
    <>
      <Button color="primary" variant="contained" onClick={() => setOpen(!open)} size="medium">
        Click Here To Open Dialog
      </Button>
      <DialogComponent
        {...args}
        className="personalInfo-dialog-component"
        color="primary"
        open={open}
        onClose={() => setOpen(false)}
        /**
         * kept this or condition just to take dynamic args from stroybook UI
         */
        dividers={true}
        title={title || 'Dialog with dividers'}
        body={
          body ||
          'Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque'
        }
      />
    </>
  );
};

export const DialogWithActions: Story<DialogComponentProps> = ({ body, title, ...args }) => {
  const onDialogueActionCallBack = () => {
    console.log('dialog action callback triggered on Save Click');
  };
  const { open, setOpen, onDialogAction } = useDialogueHook(onDialogueActionCallBack);
  return (
    <>
      <Button color="primary" variant="contained" onClick={() => setOpen(!open)} size="medium">
        Click Here To Open Dialog
      </Button>
      <DialogComponent
        {...args}
        className="personalInfo-dialog-component"
        color="primary"
        open={open}
        onClose={() => setOpen(false)}
        /**
         * kept this or condition just to take dynamic args from stroybook UI
         */
        dividers={true}
        title={title || 'Dialog with dividers'}
        body={
          body ||
          'Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque'
        }
        onAction={onDialogAction}
        actions={[
          {
            action: 'Save',
            text: 'save',
            color: 'primary',
          },
          { action: 'Cancel', text: 'cancel', color: 'secondary' },
        ]}
      />
    </>
  );
};
