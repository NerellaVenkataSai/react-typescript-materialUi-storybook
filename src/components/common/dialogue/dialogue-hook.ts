import { DialogAction } from 'components/common/dialogue/dialogue.component';
import { Dispatch, SetStateAction, useState } from 'react';

export type dialogueHookReturnType = {
  open: boolean;
  setOpen: Dispatch<SetStateAction<boolean>>;
  onDialogAction: (action: DialogAction['action'], index?: number) => void;
};

export const useDialogueHook = (callBack?: () => void): dialogueHookReturnType => {
  const [open, setOpen] = useState(false);

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const onDialogAction = (action: DialogAction['action'], index?: number) => {
    if (action === 'Cancel') {
      setOpen(false);
    }
    if (action === 'Update' || action === 'Save') {
      callBack && callBack();
    }
  };

  return {
    open,
    setOpen,
    onDialogAction,
  };
};
