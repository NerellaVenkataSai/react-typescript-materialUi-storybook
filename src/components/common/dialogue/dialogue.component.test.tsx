import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { DialogAction, DialogComponent } from './dialogue.component';

let mockOpen = true;
const mockCallBack = jest.fn();
const mockDialogueAction = jest.fn((action: DialogAction['action']) => {
  if (action === 'Cancel') {
    mockOpen = false;
  }
  if (action === 'Update') {
    mockCallBack();
  }
});
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const TestComoponent = (props: any) => (
  <DialogComponent
    className="personalInfo-dialog-component"
    color="primary"
    open={true}
    title="Dialogue Title"
    actions={[
      { action: 'Update', text: 'Update', color: 'primary' },
      { action: 'Cancel', text: 'Cancel', color: 'primary' },
    ]}
    onAction={mockDialogueAction}
    fullWidth={true}
    scroll="paper"
    dividers={false}
    {...props}
  />
);

describe('Dialogue Component', () => {
  test('Should render Dialogue Component', async () => {
    const { baseElement, getByRole, rerender } = render(<TestComoponent />);
    const dialogueComponent = baseElement.querySelector('.dialogue-component');
    const updateButton = getByRole('button', { name: 'Update' });
    const cancelButton = getByRole('button', { name: 'Cancel' });

    expect(baseElement.firstChild).toMatchSnapshot();
    expect(updateButton).toBeInTheDocument();
    expect(cancelButton).toBeInTheDocument();
    userEvent.click(updateButton);
    expect(mockDialogueAction).toBeCalledWith('Update', 0);
    expect(mockCallBack).toBeCalled();
    userEvent.click(cancelButton);
    expect(mockDialogueAction).toBeCalledWith('Cancel', 1);
    console.log(mockOpen);
    rerender(<TestComoponent open={mockOpen} />);

    await waitFor(() => {
      expect(dialogueComponent).not.toBeInTheDocument();
    });
  });
});
