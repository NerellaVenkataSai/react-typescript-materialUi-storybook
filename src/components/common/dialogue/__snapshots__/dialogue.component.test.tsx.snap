// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialogue Component Should render Dialogue Component 1`] = `
<div
  aria-hidden="true"
/>
`;
